Content.Server/AME/EntitySystems/AmeControllerSystem.cs

```diff
@@ -0,0 +1,328 @@
+    [Dependency] private readonly AppearanceSystem _appearanceSystem = default!;
+    [Dependency] private readonly ContainerSystem _containerSystem = default!;
+    [Dependency] private readonly SharedAudioSystem _audioSystem = default!;
+    [Dependency] private readonly SharedHandsSystem _handsSystem = default!;
+    [Dependency] private readonly SharedPopupSystem _popupSystem = default!;
+    [Dependency] private readonly UserInterfaceSystem _userInterfaceSystem = default!;
+
+    public override void Initialize()
+    {
+        base.Initialize();
+
+        SubscribeLocalEvent<AmeControllerComponent, ComponentStartup>(OnComponentStartup);
+        SubscribeLocalEvent<AmeControllerComponent, InteractUsingEvent>(OnInteractUsing);
+        SubscribeLocalEvent<AmeControllerComponent, PowerChangedEvent>(OnPowerChanged);
+        SubscribeLocalEvent<AmeControllerComponent, UiButtonPressedMessage>(OnUiButtonPressed);
+    }
+
+    public override void Update(float frameTime)
+    {
+        var curTime = _gameTiming.CurTime;
+        var query = EntityQueryEnumerator<AmeControllerComponent, NodeContainerComponent>();
+        while (query.MoveNext(out var uid, out var controller, out var nodes))
+        {
+            if (controller.NextUpdate <= curTime)
+                UpdateController(uid, curTime, controller, nodes);
+        }
+    }
+
+    private void UpdateController(EntityUid uid, TimeSpan curTime, AmeControllerComponent? controller = null, NodeContainerComponent? nodes = null)
+    {
+        if (!Resolve(uid, ref controller))
+            return;
+
+        controller.LastUpdate = curTime;
+        controller.NextUpdate = curTime + controller.UpdatePeriod;
+
+        if (!controller.Injecting)
+            return;
+        if (!TryGetAMENodeGroup(uid, out var group, nodes))
+            return;
+
+        if (TryComp<AmeFuelContainerComponent>(controller.JarSlot.ContainedEntity, out var fuelJar))
+        {
+            var availableInject = Math.Min(controller.InjectionAmount, fuelJar.FuelAmount);
+            var powerOutput = group.InjectFuel(availableInject, out var overloading);
+            if (TryComp<PowerSupplierComponent>(uid, out var powerOutlet))
+                powerOutlet.MaxSupply = powerOutput;
+            fuelJar.FuelAmount -= availableInject;
+            _audioSystem.PlayPvs(controller.InjectSound, uid, AudioParams.Default.WithVolume(overloading ? 10f : 0f));
+            UpdateUi(uid, controller);
+        }
+
+        controller.Stability = group.GetTotalStability();
+
+        UpdateDisplay(uid, controller.Stability, controller);
+
+        if (controller.Stability <= 0)
+            group.ExplodeCores();
+    }
+
+    public void UpdateUi(EntityUid uid, AmeControllerComponent? controller = null, ServerUserInterfaceComponent? uis = null)
+    {
+        if (!Resolve(uid, ref controller))
+            return;
+
+        if (!Resolve(uid, ref uis, logMissing: false)
```

Comment:
It delays actually generating the state to send to the ui until we are sure there's one to update. Depending on how expensive that is it might be worth it?